### PR TITLE
Thunderstore support

### PR DESCRIPTION
**Pull request by 0neGal**
Since it's currently under discussion as to whether modders should use Thunderstore to release all their mods, it's only relevant that we work on supporting it, as that's not that hard, I started working on it. It is not currently in a working state however in the devtools if you type `Browser()` it'll toggle the window for it. There's also a button under the mods section called `Find Mods` which'll open the browser.

I'll be spending most of tomorrow on working on this and getting a working version of it up and running, shouldn't take long.

<details>
<summary>Current screenshots of the WIP UI</summary>

</details>
<details>
<summary>Original screenshots of the WIP UI</summary>

</details>

##### TODO

 - [x] UI
 - [x] Search
 - [x] Localization
 - [x] Install mods
 - [x] Optimize
   - [x] Limit loading of mods to 30-50  mods
     - The user simply has to use the search to be more specific.
     - Loading too many mods at once will have quite a big performance problem.
   - [x] End of list message
     - With the prior optimization we'll have to add a message at the bottom of the browser stating max mods have been loaded, and that the user can still search to find remaining mods.
 - [x] Detection of installed mods
   - [x] Live Update
     - [x] Removing mod
     - [x] Installing mod
   - [x] Detecting already installed mod
   - [x] Detecting new updates vs just re-install
 - [x] Remove browser opening by default before merging
   - Purely because I'm lazy and didn't want to re-open it every time I reloaded the window. However I'll inevitably forget to remove it before merge if I don't have this TODO.

### CODE REVIEW

**Comment by Alystrasz** [comment]
Damn, I didn't see you were working on this.
I leave this to you, and will work on other topics.

--- 

Capture d’écran de 2022-01-20 20-47-53

**Comment by 0neGal** [comment]
@Alystrasz, if you could look at adding the missing translations that'd be nice, and further more make sure I didn't mess something up on the Windows side of things, although it should wokr just fine.

We'll merge this when French localization is merged, and #64 is also finished, along with making sure everything works on Windows as well.

**Review by AA-Delta — APPROVED** [review]

**Comment by Alystrasz** [comment]
Thunderstore mods window appears on start on Windows:

Capture d’écran 2022-02-04 235507

You can also see that window now have a white border:

Capture d’écran 2022-02-04 235351


Viper cannot be dragged anymore.
Moreover, I could install the crosshairs mod, but not the other one.

**Comment by 0neGal** [comment]
> Thunderstore mods window appears on start on Windows:

ah, that was from testing, forgot about that one. I'll remove it before merging, since we apparently have more issues that I didn't know about.

> Moreover, I could install the crosshairs mod, but not the other one.

I got a max string error, it seems the way we recursively copy files to the mods dir doesn't support files over 512mb, I'll have a look at implementing something that doesn't have this limitation. I didn't know of this limitation when I used `copy-dir`, [however it does seem like this isn't a problem](https://stackoverflow.com/questions/68230031/cannot-create-a-string-longer-than-0x1fffffe8-characters-in-json-parse) just related to `copy-dir`

> Viper cannot be dragged anymore.

This must've happened when I tried to fix not being able to click on the darkened overlay (clicking it dismisses the browser), or well, I could click on it, but only in very few chunks, and since my WM doesn't actually support window drag spaces I never noticed this :p, perhaps you could have a look at the CSS and fix it without losing the overlay feature?

> You can also see that window now have a white border:

Apparently `#bgHolder` is one pixel off the actual edge of the window... I'll fix this real quick.

**Comment by 0neGal** [comment]
All mods should now be able to be installed without problems, albeit for some reason only the crosshairs mod updates the mod list, I'll have to look into that later tomorrow.

**Comment by 0neGal** [comment]
My most recent commit should've brung back the window drag, but again I can't test this.

So all of the issues should now be fixed? I also need two more strings localized, the Info button (`"gui.browser.info": "Info"`) and the search input (`"gui.browser.search": "Search..."`) @Alystrasz @AA-Delta

**Comment by Alystrasz** [comment]
> Apparently #bgHolder is one pixel off the actual edge of the window... I'll fix this real quick.
Seems fixed.

> Viper cannot be dragged anymore.
Still cannot be dragged, I'll take a look at it.

**Comment by 0neGal** [comment]
> Still cannot be dragged, I'll take a look at it.

Thanks, do make sure the overlay can still be clicked to close when it works...

File: src/app/browser.js [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+					if (mod.includes(modname)) {
+						if (document.getElementById(modname)) {
+							Browser.setbutton(modname, string);
+						}
+					}
+				}
+			}, 1501)
+		}
+	}
+}; Browser.toggle()
```
[the remark this answers is not in the document]
Comment (0neGal) [reply src/app/browser.js:89]: This'll be removed when we do merge eventually, it's only so I don't have to keep going into the Northstar page, then mods section, then click "Find Mods" just to open the browser, I also couldn't be bothered to type it in the console every single time.

**Comment by 0neGal** [comment]
Window drag should more or less be fixed now, I tested it in XFCE where drag does work, however, I only got it to work on the titlebar area and on the `gameContainer`, I don't know if originally this wasn't how it worked, and whether it is absolute necessary that you can drag on empty space within section. In that case feel free to add support for that.

I also solved a few problems with various mod folder structures, since there's no direct standard, unfortunately. However we should have basically everything covered now, unless a modder is feeling very frisky, the only thing missing mod wise is dependency support. Which we don't need now as nobody has dependencies and we probably won't see them any time soon, however I'll add it down the line, likely submit a test package to test it all. But again, for initial release it is not important.

All that's left to do is if somebody on Windows could verify, by installing a few mods that everything works as it seems.

Note however that the Browser UI isn't always correct when it comes to guessing whether a mod is installed, if the folder name is drastically different from that of the Thunderstore package it won't detect it as being installed. So check the actual mods section for whether something installed correctly or not. I do have a way to sort of circumvent this stuff planned, and I'll likely end up getting that finished before release, I just wanted to point it out.

**Comment by Alystrasz** [comment]
Some mods documentation have been added to the wiki, maybe you can use it to deduce a correct mod's architecture, or just reference it: https://github.com/R2Northstar/NorthstarWiki/pull/43

I'll check dragging behavior on Windows tonight.

**Comment by laundmo** [comment]
> I also solved a few problems with various mod folder structures, since there's no direct standard, unfortunately. 

as the person who initiated the whole Thunderstore support, and who created the package structure: there is definitely a correct structure, its documented in the wiki, or you could take a look at https://github.com/laundmo/northstar-mod-template

Ideally any other structure should fail to install, to prevent modders from assuming their wrong structure is correct.

**Comment by laundmo** [comment]
please don't allow for other folder structures. all that does is teach modders that its okay to do whatever, instead of sticking with the official structure (which was decided together with thunderstore developers)

to my knowledge theres only 4 mods that don't use the proper structure currently, and only since i don't know who uploaded them.

**Comment by 0neGal** [comment]
> to my knowledge theres only 4 mods that don't use the proper structure currently, and only since i don't know who uploaded them.

Given this they're unlikely to change it then. I don't see a reason for not having checks in place that allow the user to still install the mod, it's for the users benefit. I don't think that outweighs the very few people who have a wrong folder structure.

And I don't see how it's promoting it, any new modder will look at the majority of mods folder structure to learn from, which'll likely be the proper one.

**Comment by Alystrasz** [comment]
> Window drag should more or less be fixed now, I tested it in XFCE where drag does work, however, I only got it to work on the titlebar area and on the gameContainer, I don't know if originally this wasn't how it worked, and whether it is absolute necessary that you can drag on empty space within section. In that case feel free to add support for that.

This does work on Windows.

**Comment by Alystrasz** [comment]
Update: minimize and close buttons cannot be clicked.

**Comment by 0neGal** [comment]
> Update: minimize and close buttons cannot be clicked.

Not only that but the section buttons also seemed borked, it is fixed now however... I'll likely merge this some time today and make a release...

**Comment by laundmo** [comment]
> And I don't see how it's promoting it, any new modder will look at the majority of mods folder structure to learn from, which'll likely be the proper one.

I have already had to message a lot of people that they should fix their mod package folder. The issue is that modder WILL NOT try to look at other packages as evident by the fact that i had to ALREADY ASK THEM TO FIX a lot of them.

The issue with trying to fix mod layouts is that its not a long-terms solution, but Thunderstore is meant to be. Long term, we want to make sure there are no mods with faulty layouts to make integration in other launchers Northstar itself  (which is something we want long-term) easier.

Please show a error message for faulty mod layouts instead of trying to fix them, as modders will definitely use Viper as a way to test if their mods work.

**Comment by 0neGal** [comment]
> Please show a error message for faulty mod layouts instead of trying to fix them, as modders will definitely use Viper as a way to test if their mods work.

I was already planning on doing this, and overall I was planning on having a tiny toast popup on install etc, however since it wasn't necessarily related to Thunderstore support I was gonna add it after merging (rn)

**Comment by 0neGal** [comment]
@laundmo would you be able to find a mod that has the wrong folder structure, all the ones that used to have them seem to be fixed (I assume thanks to you), so I don't have any that I can really test it on.

**Comment by laundmo** [comment]
https://northstar.thunderstore.io/package/MILE/pilotsword/ and all of the other mods this modder uploaded

**Pull request merged**